File: Source/UI/Frames/FrameManager.cs
using System;
using System.Collections.Generic;
using WCSharp.Api;
using static WCSharp.Api.Common;

public static class FrameManager
{
    public static trigger StatsTrigger = trigger.Create();
    public static trigger ShopTrigger = trigger.Create();
    public static trigger RewardsTrigger = trigger.Create();
    public static framehandle MusicButton;
    public static framehandle ShopButton;
    public static framehandle RewardsButton;
    public static framehandle Backdrop;

    private const float ButtonWidth = 0.053f;
    private const float ButtonHeight = 0.028f;

    private static framehandle GameUI = originframetype.GameUI.GetOriginFrame(0);
    private static trigger ESCTrigger = trigger.Create();
    private static string TEXT_COLOR = Colors.COLOR_YELLOW;
    private static string HOTKEY_COLOR = Colors.COLOR_YELLOW_ORANGE;

    private static readonly Action _cachedUIPosition = RepositionBackdropAction();
    private static List<framehandle> _frames = new List<framehandle>();

    public static void Initialize()
    {
        try
        {
            BlzLoadTOCFile("war3mapImported\\templates.toc");
            RemoveUnwantedFrames();
            ButtonsBackdrop();
            CreateRewardsButton();
            CreateMusicButton();
            CreateShopButton();
            MusicFrame.Initialize();
            Utility.SimpleTimer(1.0f, ESCHideFrames);
        }
        catch (Exception ex)
        {
            Logger.Critical($"Error in FrameManager.Initialize: {ex.Message}");
            throw;
        }
    }

    public static void InitAllFrames()
    {
        ShopFrame.Initialize();
        RewardsFrame.Initialize();
        InitalizeButtons();
        InitFramesList();
    }

    public static void InitalizeButtons()
    {
        Backdrop.Visible = true;
        RewardsButton.Visible = true;
        MusicButton.Visible = true;
        ShopButton.Visible = true;
    }

    public static framehandle CreateHeaderFrame(framehandle parent)
    {
        var header = framehandle.Create("BACKDROP", $"{parent.Name}Header", parent, "QuestButtonDisabledBackdropTemplate", 0);
        var width = parent.Width;
        var height = 0.0225f;
        header.SetPoint(framepointtype.TopLeft, 0, 0.0125f, parent, framepointtype.TopLeft);
        header.SetSize(width, height);

        var title = framehandle.Create("TEXT", $"{parent.Name}Title", header, "ScriptDialogText", 0);
        title.SetPoint(framepointtype.Center, 0, 0, header, framepointtype.Center);
        title.SetSize(width, height);
        title.Text = $"{Colors.COLOR_YELLOW}{parent.Name}{Colors.COLOR_RESET}";
        title.SetTextAlignment(textaligntype.Center, textaligntype.Center);

        var closeButton = framehandle.Create("GLUETEXTBUTTON", $"{parent.Name}CloseButton", header, "ScriptDialogButton", 0);
        closeButton.SetPoint(framepointtype.TopRight, -0.0025f, -0.0025f, header, framepointtype.TopRight);
        closeButton.SetSize(height - 0.005f, height - 0.005f);
        closeButton.Text = "X";

        // Close Actions
        var closeTrigger = trigger.Create();
        closeTrigger.RegisterFrameEvent(closeButton, frameeventtype.Click);
        closeTrigger.AddAction(ErrorHandler.Wrap(() =>
        {
            if (!@event.Player.IsLocal) return;
            parent.Visible = false;
        }));

        return header;
    }

    private static void RemoveUnwantedFrames()
    {
        var resourceBarText = BlzGetFrameByName("ResourceBarSupplyText", 0);
        BlzFrameGetChild(BlzFrameGetChild(GameUI, 5), 0);
        resourceBarText.Text = "0:00";
        //timeDayDisplay.Visible = false;
    }

    private static void InitFramesList()
    {
        _frames.Add(ShopFrame.shopFrame);
        _frames.Add(RewardsFrame.RewardFrame);
        _frames.Add(MusicFrame.MusicFramehandle);
    }

    private static void CreateRewardsButton()
    {
        RewardsButton = framehandle.Create("GLUETEXTBUTTON", "RewardsButton", Backdrop, "ScriptDialogButton", 0);
        RewardsButton.SetPoint(framepointtype.Center, 0, 0, Backdrop, framepointtype.Center);
        RewardsButton.SetSize(ButtonWidth, ButtonHeight);
        var shopText = framehandle.Create("TEXT", "RewardsText", RewardsButton, "", 0);
        shopText.Text = $"{TEXT_COLOR}Rewards{HOTKEY_COLOR}(-)|r";
        shopText.SetPoint(framepointtype.Center, 0, 0, RewardsButton, framepointtype.Center);
        shopText.SetScale(0.9f);
        shopText.Enabled = false;
        RewardsTrigger.RegisterFrameEvent(RewardsButton, frameeventtype.Click);
        RewardsTrigger.AddAction(ErrorHandler.Wrap(RewardsFrame.RewardsFrameActions));
        RewardsButton.Visible = false;
    }

    private static void CreateMusicButton()
    {
        MusicButton = framehandle.Create("GLUETEXTBUTTON", "MusicButton", Backdrop, "ScriptDialogButton", 0);
        MusicButton.SetPoint(framepointtype.TopRight, 0, 0, RewardsButton, framepointtype.TopLeft);
        MusicButton.SetSize(ButtonWidth, ButtonHeight);
        var shopText = framehandle.Create("TEXT", "MusicText", MusicButton, "", 0);
        shopText.Text = $"{TEXT_COLOR}Music{HOTKEY_COLOR}(0)";
        shopText.SetPoint(framepointtype.Center, 0, 0, MusicButton, framepointtype.Center);
        shopText.SetScale(0.98f);
        shopText.Enabled = false;
        StatsTrigger.RegisterFrameEvent(MusicButton, frameeventtype.Click);
        StatsTrigger.AddAction(ErrorHandler.Wrap(MusicFrame.MusicFrameActions));
        MusicButton.Visible = false;
    }

    private static void CreateShopButton()
    {
        ShopButton = framehandle.Create("GLUETEXTBUTTON", "ShopButton", Backdrop, "ScriptDialogButton", 0);
        ShopButton.SetPoint(framepointtype.TopLeft, 0, 0, RewardsButton, framepointtype.TopRight);
        ShopButton.SetSize(ButtonWidth, ButtonHeight);
        var shopText = framehandle.Create("TEXT", "ShopText", ShopButton, "", 0);
        shopText.Text = $"{TEXT_COLOR}Shop{HOTKEY_COLOR}(=)";
        shopText.SetPoint(framepointtype.Center, 0, 0, ShopButton, framepointtype.Center);
        shopText.SetScale(1.0f);
        shopText.Enabled = false;
        ShopTrigger.RegisterFrameEvent(ShopButton, frameeventtype.Click);
        ShopTrigger.AddAction(ErrorHandler.Wrap(ShopFrame.ShopFrameActions));
        ShopButton.Visible = false;
    }

    private static void ButtonsBackdrop()
    {
        Backdrop = framehandle.Create("BACKDROP", "ButtonsBackdrop", GameUI, "QuestButtonDisabledBackdropTemplate", 0);
        BlzGetFrameByName("ResourceBarGoldText", 0);
        Backdrop.SetPoint(framepointtype.Top, 0, 0, GameUI, framepointtype.Top);
        Backdrop.SetSize(0.16f, 0.035f);
        Backdrop.SetScale(1.0f);
        Backdrop.Visible = false;
        RepositionBackdrop();
    }

    private static void RepositionBackdrop()
    {
        var t = timer.Create();
        var nameFrame = BlzGetFrameByName("ConsoleUIBackdrop", 0);

        t.Start(1.0f, true, _cachedUIPosition);
    }

    private static Action RepositionBackdropAction()
    {
        return () => {
            try
            {
                var nameFrame = BlzGetFrameByName("ConsoleUIBackdrop", 0);
                var x = nameFrame.Width / 4;
                var h = nameFrame.Height / 8;
                var yOffSet = nameFrame.Height / 8;
                Backdrop.SetPoint(framepointtype.Top, 0, yOffSet, nameFrame, framepointtype.Top);
                Backdrop.SetSize(x, h);
            }
            catch (Exception e)
            {
                Logger.Critical($"Error in RepositionBackdropAction: {e.Message}");
            }
        };
    }

    private static void ESCHideFrames()
    {
        for (int i = 0; i < Globals.ALL_PLAYERS.Count; i++)
        {
            var player = Globals.ALL_PLAYERS[i];
            ESCTrigger.RegisterPlayerEvent(player, playerevent.EndCinematic);
        }
        ESCTrigger.AddAction(ErrorHandler.Wrap(ESCActions));
    }

    public static void RefreshFrame(framehandle frame)
    {
        frame.Visible = !frame.Visible;
        frame.Visible = !frame.Visible;
    }

    private static void ESCActions()
    {
        var player = @event.Player;
        if (!player.IsLocal) return;
        RewardsFrame.RewardFrame.Visible = false;
        ShopFrame.shopFrame.Visible = false;
        MusicFrame.MusicFramehandle.Visible = false;
    }

    public static void HideOtherFrames(framehandle currentFrame)
    {
        for(int i = 0; i < _frames.Count; i++)
        {
            if (_frames[i] != currentFrame) _frames[i].Visible = false;
        }
    }
}

File: src/UI/Frames/FrameManager.ts
import { Logger } from 'src/Events/Logger/Logger'
import { Globals } from 'src/Global/Globals'
import { Colors } from 'src/Utility/Colors/Colors'
import { Action } from 'src/Utility/CSUtils'
import { ErrorHandler } from 'src/Utility/ErrorHandler'
import { Utility } from 'src/Utility/Utility'
import { blzCreateFrameByType, blzGetFrameByName, getTriggerPlayer } from 'src/Utility/w3tsUtils'
import { Frame, Timer, Trigger } from 'w3ts'
import { MusicFrame } from './MusicFrame'
import { RewardsFrame } from './RewardsFrame'
import { ShopFrame } from './ShopFrame'

export class FrameManager {
    public static StatsTrigger: Trigger = Trigger.create()!
    public static ShopTrigger: Trigger = Trigger.create()!
    public static RewardsTrigger: Trigger = Trigger.create()!
    public static MusicButton: Frame
    public static ShopButton: Frame
    public static RewardsButton: Frame
    public static Backdrop: Frame

    private static ButtonWidth = 0.053
    private static ButtonHeight = 0.028

    private static GameUI: Frame
    private static ESCTrigger: Trigger = Trigger.create()!
    private static TEXT_COLOR: string = Colors.COLOR_YELLOW
    private static HOTKEY_COLOR: string = Colors.COLOR_YELLOW_ORANGE

    private static _cachedUIPosition: Action
    private static _frames: Frame[] = []

    public static Initialize() {
        try {
            this.GameUI = Frame.fromHandle(BlzGetOriginFrame(ORIGIN_FRAME_GAME_UI, 0))!
            this._cachedUIPosition = FrameManager.RepositionBackdropAction()
            BlzLoadTOCFile('war3mapImported\\templates.toc')
            FrameManager.RemoveUnwantedFrames()
            FrameManager.ButtonsBackdrop()
            FrameManager.CreateRewardsButton()
            FrameManager.CreateMusicButton()
            FrameManager.CreateShopButton()
            Utility.SimpleTimer(1.0, FrameManager.ESCHideFrames)
        } catch (ex: any) {
            Logger.Critical(`Error in FrameManager.Initialize: ${ex}`)
            throw ex
        }
    }

    public static InitalizeButtons() {
        FrameManager.Backdrop.visible = true
        FrameManager.RewardsButton.visible = true
        FrameManager.MusicButton.visible = true
        FrameManager.ShopButton.visible = true
    }

    public static CreateHeaderFrame(parent: Frame): Frame {
        let header = blzCreateFrameByType(
            'BACKDROP',
            `${parent.getName()}Header`,
            parent,
            'QuestButtonDisabledBackdropTemplate',
            0
        )
        let width = parent.width
        let height = 0.0225
        header.setPoint(FRAMEPOINT_TOPLEFT, parent, FRAMEPOINT_TOPLEFT, 0, 0.0125)
        header.setSize(width, height)

        let title = blzCreateFrameByType('TEXT', `${parent.getName()}Title`, header, 'ScriptDialogText', 0)
        title.setPoint(FRAMEPOINT_CENTER, header, FRAMEPOINT_CENTER, 0, 0)
        title.setSize(width, height)
        title.text = `${Colors.COLOR_YELLOW}${parent.getName()}${Colors.COLOR_RESET}`
        BlzFrameSetTextAlignment(title.handle, TEXT_JUSTIFY_CENTER, TEXT_JUSTIFY_CENTER)

        let closeButton = blzCreateFrameByType(
            'GLUETEXTBUTTON',
            `${parent.getName()}CloseButton`,
            header,
            'ScriptDialogButton',
            0
        )
        closeButton.setPoint(FRAMEPOINT_TOPRIGHT, header, FRAMEPOINT_TOPRIGHT, -0.0025, -0.0025)
        closeButton.setSize(height - 0.005, height - 0.005)
        closeButton.text = 'X'

        // Close Actions
        let closeTrigger = Trigger.create()!
        closeTrigger.triggerRegisterFrameEvent(closeButton, FRAMEEVENT_CONTROL_CLICK)
        closeTrigger.addAction(() => {
            if (!getTriggerPlayer().isLocal()) return
            parent.visible = false
        })

        return header
    }

    private static RemoveUnwantedFrames() {
        let resourceBarText = blzGetFrameByName('ResourceBarSupplyText', 0)
        BlzFrameGetChild(BlzFrameGetChild(FrameManager.GameUI.handle, 5)!, 0)
        resourceBarText.text = '0:00'
        //timeDayDisplay.visible = false;
    }

    public static InitFramesList() {
        FrameManager._frames.push(ShopFrame.shopFrame)
        FrameManager._frames.push(RewardsFrame.RewardFrame)
        FrameManager._frames.push(MusicFrame.MusicFramehandle)
    }

    private static CreateRewardsButton() {
        FrameManager.RewardsButton = blzCreateFrameByType(
            'GLUETEXTBUTTON',
            'RewardsButton',
            FrameManager.Backdrop,
            'ScriptDialogButton',
            0
        )
        FrameManager.RewardsButton.setPoint(FRAMEPOINT_CENTER, FrameManager.Backdrop, FRAMEPOINT_CENTER, 0, 0)
        FrameManager.RewardsButton.setSize(this.ButtonWidth, this.ButtonHeight)
        let shopText = blzCreateFrameByType('TEXT', 'RewardsText', FrameManager.RewardsButton, '', 0)
        shopText.text = `${Colors.COLOR_YELLOW}Rewards${Colors.COLOR_LAVENDER}(-)|r`
        shopText.setPoint(FRAMEPOINT_CENTER, FrameManager.RewardsButton, FRAMEPOINT_CENTER, 0, 0)
        shopText.setScale(0.9)
        shopText.enabled = false
        FrameManager.RewardsTrigger.triggerRegisterFrameEvent(FrameManager.RewardsButton, FRAMEEVENT_CONTROL_CLICK)
        FrameManager.RewardsTrigger.addAction(ErrorHandler.Wrap(() => RewardsFrame.RewardsFrameActions()))
        FrameManager.RewardsButton.visible = false
    }

    private static CreateMusicButton() {
        FrameManager.MusicButton = blzCreateFrameByType(
            'GLUETEXTBUTTON',
            'MusicButton',
            FrameManager.Backdrop,
            'ScriptDialogButton',
            0
        )
        FrameManager.MusicButton.setPoint(FRAMEPOINT_TOPRIGHT, FrameManager.RewardsButton, FRAMEPOINT_TOPLEFT, 0, 0)
        FrameManager.MusicButton.setSize(this.ButtonWidth, this.ButtonHeight)
        let shopText = blzCreateFrameByType('TEXT', 'MusicText', FrameManager.MusicButton, '', 0)
        shopText.text = `${Colors.COLOR_YELLOW}Music${Colors.COLOR_LAVENDER}(0)`
        shopText.setPoint(FRAMEPOINT_CENTER, FrameManager.MusicButton, FRAMEPOINT_CENTER, 0, 0)
        shopText.setScale(0.98)
        shopText.enabled = false
        FrameManager.StatsTrigger.triggerRegisterFrameEvent(FrameManager.MusicButton, FRAMEEVENT_CONTROL_CLICK)
        FrameManager.StatsTrigger.addAction(ErrorHandler.Wrap(() => MusicFrame.MusicFrameActions()))
        FrameManager.MusicButton.visible = false
    }

    private static CreateShopButton() {
        FrameManager.ShopButton = blzCreateFrameByType(
            'GLUETEXTBUTTON',
            'ShopButton',
            FrameManager.Backdrop,
            'ScriptDialogButton',
            0
        )
        FrameManager.ShopButton.setPoint(FRAMEPOINT_TOPLEFT, FrameManager.RewardsButton, FRAMEPOINT_TOPRIGHT, 0, 0)
        FrameManager.ShopButton.setSize(this.ButtonWidth, this.ButtonHeight)
        let shopText = blzCreateFrameByType('TEXT', 'ShopText', FrameManager.ShopButton, '', 0)
        shopText.text = `${Colors.COLOR_YELLOW}Shop${Colors.COLOR_LAVENDER}(=)`
        shopText.setPoint(FRAMEPOINT_CENTER, FrameManager.ShopButton, FRAMEPOINT_CENTER, 0, 0)
        shopText.setScale(1.0)
        shopText.enabled = false
        FrameManager.ShopTrigger.triggerRegisterFrameEvent(FrameManager.ShopButton, FRAMEEVENT_CONTROL_CLICK)
        FrameManager.ShopTrigger.addAction(ErrorHandler.Wrap(() => ShopFrame.ShopFrameActions()))
        FrameManager.ShopButton.visible = false
    }

    private static ButtonsBackdrop() {
        FrameManager.Backdrop = blzCreateFrameByType(
            'BACKDROP',
            'ButtonsBackdrop',
            FrameManager.GameUI,
            'QuestButtonDisabledBackdropTemplate',
            0
        )
        blzGetFrameByName('ResourceBarGoldText', 0)
        FrameManager.Backdrop.setPoint(FRAMEPOINT_TOP, FrameManager.GameUI, FRAMEPOINT_TOP, 0, 0)
        FrameManager.Backdrop.setSize(0.16, 0.035)
        FrameManager.Backdrop.setScale(1.0)
        FrameManager.Backdrop.visible = false
        FrameManager.RepositionBackdrop()
    }

    private static RepositionBackdrop() {
        let t = Timer.create()
        let nameFrame = blzGetFrameByName('ConsoleUIBackdrop', 0)

        t.start(1.0, true, FrameManager._cachedUIPosition)
    }

    private static RepositionBackdropAction(): Action {
        return () => {
            try {
                let nameFrame = blzGetFrameByName('ConsoleUIBackdrop', 0)
                let x = nameFrame.width / 4
                let h = nameFrame.height / 8
                let yOffSet = nameFrame.height / 8
                FrameManager.Backdrop.setPoint(FRAMEPOINT_TOP, nameFrame, FRAMEPOINT_TOP, 0, yOffSet)
                FrameManager.Backdrop.setSize(x, h)
            } catch (e: any) {
                Logger.Critical(`Error in RepositionBackdropAction: ${e}`)
            }
        }
    }

    private static ESCHideFrames() {
        for (let i = 0; i < Globals.ALL_PLAYERS.length; i++) {
            let player = Globals.ALL_PLAYERS[i]
            FrameManager.ESCTrigger.registerPlayerEvent(player, EVENT_PLAYER_END_CINEMATIC)
        }
        FrameManager.ESCTrigger.addAction(() => FrameManager.ESCActions())
    }

    public static RefreshFrame(frame: Frame) {
        frame.visible = !frame.visible
        frame.visible = !frame.visible
    }

    private static ESCActions() {
        let player = getTriggerPlayer()
        if (!player.isLocal()) return
        RewardsFrame.RewardFrame.visible = false
        ShopFrame.shopFrame.visible = false
        MusicFrame.MusicFramehandle.visible = false
    }

    public static HideOtherFrames(currentFrame: Frame) {
        for (let i = 0; i < FrameManager._frames.length; i++) {
            if (FrameManager._frames[i] !== currentFrame) FrameManager._frames[i].visible = false
        }
    }
}
